src/algorithms/semihdp_sampler.hpp

```diff
@@ -0,0 +1,205 @@
+#ifndef SRC_ALGORITHMS_SEMI_HDP_SAMPLER_HPP
+#define SRC_ALGORITHMS_SEMI_HDP_SAMPLER_HPP
+
+#include <omp.h>
+#include <proto/cpp/semihdp.pb.h>
+
+#include <Eigen/Dense>
+#include <numeric>
+#include <src/collectors/memory_collector.hpp>
+#include <src/hierarchies/base_hierarchy.hpp>
+#include <src/utils/distributions.hpp>
+#include <src/utils/rng.hpp>
+#include <stan/math/prim.hpp>
+#include <stdexcept>
+#include <vector>
+
+using bayesmix::MarginalState;
+using bayesmix::SemiHdpState;
+using Eigen::MatrixXd;
+using Eigen::VectorXd;
+
+/*
+ * This class implements the algorithm for posterior simulation under the
+ * semi-hierarchical Dirichlet process in [1].
+ * 
+ * Extra goodies: we can now tune the pseudo-prior generation, by randomly
+ * drawing the cardinalities from a multinomial distribution, and perturbing
+ * the atoms of the mixing measure.
+ *
+ * [1]: "The semi-hierarchical Dirichlet Process and its application to
+ * clustering homogeneous distributions", Beraha, Guglielmi and Quintana
+ * arXiv: 2005.10287
+ */
+
+class SemiHdpSampler {
+ protected:
+  bayesmix::SemiHdpParams params;
+
+  std::vector<MatrixXd> data;  // one vector per group
+  int ngroups;
+  std::vector<int> n_by_group;
+
+  std::shared_ptr<BaseHierarchy> G0_master_hierarchy;
+  std::shared_ptr<BaseHierarchy> G00_master_hierarchy;
+
+  // these should be just copies
+  std::vector<std::vector<std::shared_ptr<BaseHierarchy>>> theta_star;
+  // idiosincratic tables
+  std::vector<std::vector<std::shared_ptr<BaseHierarchy>>> theta_tilde;
+  // shared tables
+  std::vector<std::shared_ptr<BaseHierarchy>> taus;
+  // counts
+  std::vector<std::vector<int>> n_by_theta_star;
+
+  // stuff for pseudoprior
+  std::vector<std::vector<std::shared_ptr<BaseHierarchy>>> theta_star_pseudo;
+  std::vector<std::vector<int>> n_by_theta_star_pseudo;
+
+  // theta_{ij} = theta_{c_i, s_{ij}}
+  std::vector<std::vector<int>> s;
+  // which restaurant each grups enters from
+  std::vector<int> c;
+  std::vector<bool> is_used_c;
+  VectorXd omega;
+  std::vector<std::vector<int>> t, v;
+
+  // number of theta_stars equal to tau_h across all groups
+  std::vector<int> m;
+
+  double w = 0.5;
+  double alpha, gamma, a_w, b_w;
+
+  std::vector<MemoryCollector<MarginalState>> pseudoprior_collectors;
+  int pseudo_iter;
+
+  bool adapt = false;
+  std::string c_update;
+
+ public:
+  SemiHdpSampler() {}
+  ~SemiHdpSampler() {}
+
+  SemiHdpSampler(const std::vector<MatrixXd> &data,
+                 std::shared_ptr<BaseHierarchy> hier,
+                 bayesmix::SemiHdpParams params);
```

Comment:
If you want this class to be able to be chosen at runtime, for simplicity's sake, you should use invidiual setters rather than a constructor to pass these arguments (thus this constructor would be useless)

Reply:
at the moment this is the only sampler dealing with multiple groups of data, so I would keep it as it is.
If in the future we want to add more samplers of these kinds (such as HDP, Nested Dirichlet Process ecc...) I will change this!